hudi-common/src/test/java/org/apache/hudi/common/functional/TestHoodieLogFormat.java

```diff
@@ -671,11 +658,188 @@ public void testBasicAppendAndScanMultipleFiles(ExternalSpillableMap.DiskMapType
+                                                             boolean isCompressionEnabled,
+                                                             boolean readBlocksLazily,
+                                                             boolean useScanV2)
+      throws IOException, URISyntaxException, InterruptedException {
+    // Generate 3 delta-log files w/ random records
+    Schema schema = HoodieAvroUtils.addMetadataFields(getSimpleSchema());
+    List<IndexedRecord> genRecords = SchemaTestUtil.generateHoodieTestRecords(0, 300);
+
+    Set<HoodieLogFile> logFiles = writeLogFiles(partitionPath, schema, genRecords, 3);
+
+    FileCreateUtils.createDeltaCommit(basePath, "100", fs);
+
+    // scan all log blocks (across multiple log files)
+    HoodieMergedLogRecordScanner scanner = HoodieMergedLogRecordScanner.newBuilder()
+        .withFileSystem(fs)
+        .withBasePath(basePath)
+        .withLogFilePaths(
+            logFiles.stream()
+                .map(logFile -> logFile.getPath().toString()).collect(Collectors.toList()))
+        .withReaderSchema(schema)
+        .withLatestInstantTime("100")
+        .withMaxMemorySizeInBytes(1024L)
+        .withReadBlocksLazily(readBlocksLazily)
+        .withReverseReader(false)
+        .withBufferSize(BUFFER_SIZE)
+        .withSpillableMapBasePath(spillableBasePath)
+        .withDiskMapType(diskMapType)
+        .withBitCaskDiskMapCompressionEnabled(isCompressionEnabled)
+        .withUseScanV2(useScanV2)
+        .withRecordMerger(HoodieRecordUtils.loadRecordMerger(HoodieAvroRecordMerger.class.getName()))
+        .withForceFullScan(false)
+        .build();
+
+
+    List<String> sampledRecordKeys = Arrays.asList(
+        "00509b14-3d1a-4283-9a8c-c72b971a9d06",
+        "006b2f57-9bf7-4634-910c-c91542ea61e5",
+        "007fc45d-7ce2-45be-8765-0b9082412518",
+        "00826e50-73b4-4cb0-9d5a-375554d5e0f7"
+    );
+
+    List<IndexedRecord> sampledRecords = genRecords.stream()
+        .filter(r -> sampledRecordKeys.contains(((GenericRecord) r).get(HoodieRecord.RECORD_KEY_METADATA_FIELD).toString()))
+        .collect(Collectors.toList());
+
+    List<String> sampledKeyPrefixes = Collections.singletonList("00");
+
+    //
+    // Step 1: Scan by a list of keys
+    //
+
+    scanner.scanByKeyPrefixes(sampledKeyPrefixes);
+
+    List<HoodieRecord> scannedHoodieRecords = new ArrayList<>();
+    List<IndexedRecord> scannedAvroRecords = new ArrayList<>();
+    for (HoodieRecord record : scanner) {
+      scannedHoodieRecords.add(record);
+      scannedAvroRecords.add((IndexedRecord)
+          ((HoodieAvroRecord) record).getData().getInsertValue(schema).get());
+    }
+
+    //assertEquals(sort(sampledRecords), sort(scannedAvroRecords));
+
+    //
+    // Step 2: Scan by the same list of keys (no new scanning should be performed,
+    //         in this case, and same _objects_ have to be returned)
+    //
+
+    scanner.scanByKeyPrefixes(sampledKeyPrefixes);
```

Comment:
is it possible to test this flow.
in call1, we do scanByKeyPrefixes("00","01")
and in call2, we do  scanByKeyPrefixes("01","02","03")
and in call2, we do scanByKeyPrefixes("00","02")

also, how can we ensure call2 did not invoke actual scanning internally ? should we mock and verify the times an internal method is invoked?

Reply:
Discussed offline: existing scenario covers exactly the branches, and there's no branch coverage increase with this adjustment